--- src/java.base/unix/native/libjava/io_util_md.c.orig	2026-04-08 16:09:49.970437852 +0200
+++ src/java.base/unix/native/libjava/io_util_md.c	2026-04-08 16:10:12.180121302 +0200
@@ -34,6 +34,10 @@
 #include <sys/ioctl.h>
 #endif
 
+#if defined(__sun)
+#include <sys/filio.h>
+#endif
+
 #if defined(__linux__)
 #include <linux/fs.h>
 #include <sys/stat.h>
--- src/java.base/unix/native/libnio/ch/NativeThread.c.orig
+++ src/java.base/unix/native/libnio/ch/NativeThread.c
@@ -44,6 +44,9 @@
 #elif defined(_ALLBSD_SOURCE)
   /* Also defined in net/bsd_close.c */
   #define INTERRUPT_SIGNAL SIGIO
+#elif defined(__sun)
+  /* Solaris/Illumos - use SIGUSR2 for interrupt signal */
+  #define INTERRUPT_SIGNAL SIGUSR2
 #else
   #error "missing platform-specific definition here"
 #endif
--- src/java.base/unix/native/libnio/ch/Net.c.orig      2026-04-08 16:14:13.413803096 +0200
+++ src/java.base/unix/native/libnio/ch/Net.c   2026-04-08 16:14:52.821792093 +0200
@@ -47,6 +47,11 @@
 #include <sys/utsname.h>
 #endif

+#if defined(__sun)
+#include <sys/filio.h>
+#endif
+
+
 /**
  * IP_MULTICAST_ALL supported since 2.6.31 but may not be available at
  * build time.
--- src/java.base/unix/native/libnio/ch/UnixFileDispatcherImpl.c.orig   2026-04-08 18:12:35.712325951 +0200
+++ src/java.base/unix/native/libnio/ch/UnixFileDispatcherImpl.c        2026-04-08 18:14:16.403831106 +0200
@@ -38,6 +38,10 @@
 #include <sys/ioctl.h>
 #endif

+#if defined(__solaris__) || defined(__sun)
+#include <sys/filio.h>
+#endif
+
 #include "jni.h"
 #include "nio.h"
 #include "nio_util.h"
